Retry InitializeEnclave on Windows on SGX EINIT SGX_UNMASKED_EVENT error

anitagov opened the pull request:
Fixes bug (Issue #1794 ) seen with Windows CI where a test fails during the call to Windows' InitializeEnclave() system call with the underlying SGX EINIT error SGX_UNMASKED_EVENT (0x80). Mitigating this with retry logic at the OE SDK level as inhibiting the unmasked events would require Windows kernel support.

anitagov commented:
Couple of important comments:
1) Did not add any delay or sleep as the context switch to kernel mode should suffice for the delay.
2) Set the MAX_RETRY_COUNT to 50.
3) Did not see any failure with 4 consecutive CI runs. So, would like this to get into master.
We an tweak 1) and/or 2) if needed if we see failures with Windows CI.

This change does not affect Linux.

jxyang commented on host/sgx/sgxload.c:
```diff
@@ -545,24 +546,37 @@ oe_result_t oe_sgx_create_enclave(
+
+            if (!base)
+            {
+                if (enclave_error == SGX_UNMASKED_EVENT)
```
We are not going to see an error message in the worst case, we failed to create the enclave with after retrying MAX_EINIT_RETRY_COUNT times.

anitagov replied:
Good point. Will incorporate

CodeMonkeyLeet replied:
Your fix still doesn't address the comment but instead spews an error message on every retry, which is not meaningful. I'm not sure why this loop needs to be so complicated even.

```c
void* base = NULL;
uint32_t enclave_error;

for (DWORD i = 0; i < MAX_EINIT_RETRY_COUNT; i++)
    base = enclave_create(
        NULL, /* Let OS choose the enclave base address */
        secs->size,
        secs->size,
        ENCLAVE_TYPE_SGX1,
        (const void*)secs,
        sizeof(sgx_secs_t),
        &enclave_error);
        
	if (enclave_error != SGX_UNMASKED_EVENT)
	    break;
}

if (!base)
{
    OE_RAISE_MSG(
        OE_PLATFORM_ERROR,
	"CreateEnclave failed (err=%#x)",
        enclave_error);
}

secs->base = (uint64_t)base;
```

anitagov replied:
> Your fix still doesn't address the comment but instead spews an error message on every retry, which is not meaningful. I'm not sure why this loop needs to be so complicated even.
 
The error message is printed only when the retry max count is reached. Otherwise, it continues to the while loop:

                    if (einit_retry_count == MAX_EINIT_RETRY_COUNT)
                    {
                        OE_RAISE_MSG(
                            OE_PLATFORM_ERROR,
                            "CreateEnclave failed with SGX_UNMASKED_EVENT "
                            "after %d retries",
                            MAX_EINIT_RETRY_COUNT);
                    }
                    else
                        continue;
             }
OE_RAISE_MSG not only prints the error message but will jump to done and exit the routine.

I prefer for loops instead of while too.

anitagov replied:
> Your fix still doesn't address the comment but instead spews an error message on every retry, which is not meaningful. I'm not sure why this loop needs to be so complicated even.
> 
> ```c
> void* base = NULL;
> uint32_t enclave_error;
> 
> for (DWORD i = 0; i < MAX_EINIT_RETRY_COUNT; i++)
> {
>     base = enclave_create(
>         NULL, /* Let OS choose the enclave base address */
>         secs->size,
>         secs->size,
>         ENCLAVE_TYPE_SGX1,
>         (const void*)secs,
>         sizeof(sgx_secs_t),
>         &enclave_error);
>         
>     if (enclave_error != SGX_UNMASKED_EVENT)
>         break;
> }
> 
> if (!base)
> {
>     OE_RAISE_MSG(
>         OE_PLATFORM_ERROR,
> 	"CreateEnclave failed (err=%#x)",
>         enclave_error);
> }
> 
> secs->base = (uint64_t)base;
> ```

@CodeMonkeyLeet - I did not modify the section that you reference. I only modified the 
#elif defined(_WIN32) section with the create_enclave function. This is the section that gets run with the current AESM flow on Windows.

anitagov replied:
We can add the retry if needed to this section once we have support for Windows CI with DCAP and hit this issue. This is the reason that I did not make any changes to this section.

anitagov commented:
bors try

CodeMonkeyLeet commented on host/sgx/sgxload.c:
```diff
@@ -545,24 +546,46 @@ oe_result_t oe_sgx_create_enclave(
-            sizeof(ENCLAVE_CREATE_INFO_SGX),
-            &enclave_error);
+        DWORD einit_retry_count = 0;
+        while (einit_retry_count < MAX_EINIT_RETRY_COUNT)
```
There should be a comment to indicate why we only do this retry loop on Windows and not Linux. You can lift text from #1794. You should also indicate why we think a sleep between loops is unnecessary (i.e. cost of kernel transition, as stated in the PR comments).

anitagov replied:
ok, will incorporate.

jxyang commented:
@anitagov The failure happens at `InitializeEnclave`, not `CreateEnclave`. Sorry I may have misled you. It's in the same file, but in a different function `oe_sgx_initialize_enclave`.  I'd write the loop this way:
```c
        int i = 0;
        bool result = false;
        do
        {
            result = InitializeEnclave(
                    GetCurrentProcess(),
                    (LPVOID)addr,
                    &info,
                    sizeof(info),
                    &enclave_error);
        } while (i++ < MAX_EINIT_RETRY_COUNT &&
                 enclave_error == SGX_UNMASKED_EVENT);

        if (!result)
            OE_RAISE_MSG(
                OE_PLATFORM_ERROR,
                "InitializeEnclave failed (err=%#x)",
                enclave_error);
```

anitagov commented:
@jxyang - Yes, your comments misled me. I would have looked at the failing signature in the issue/failing pipelines to figure out the fix :-)

anitagov commented:
@jxyang @CodeMonkeyLeet - I fixed InitializeEnclave, the correct path right now. Added comments and have a cleaner flow. Please review.

Currently testing in sandbox mode. I will squash the commits before merge to master.

jxyang commented:
LGTM

anitagov commented:
Got 4 clean sandbox runs in a row. Waiting for approval.

CodeMonkeyLeet commented on host/sgx/sgxload.c:
```diff
@@ -918,16 +919,39 @@ oe_result_t oe_sgx_initialize_enclave(
             sizeof(info.EInitToken)));
 
-        if (!InitializeEnclave(
+        bool result = false;
```
Check that compilation on Windows doesn't complain about the implicit cast from BOOL to bool. Probably just easier to use the Windows type. 

You're also shadowing the existing `result` variable in this function, please rename it:

```c
BOOL success = FALSE;
```

anitagov replied:
ok

anitagov commented:
bors r+

anitagov commented:
@yakman2020 - Nice if this gets into Release 0.6 although it only affects Windows (CI).

CodeMonkeyLeet commented on host/sgx/sgxload.c:
```diff
@@ -918,16 +919,39 @@ oe_result_t oe_sgx_initialize_enclave(
             sizeof(info.EInitToken)));
 
-        if (!InitializeEnclave(
+        BOOL success = false;
```
You're still doing an implicit cast via the assignment of `false` to a BOOL type.

anitagov replied:
Changed to FALSE.

anitagov commented:
bors r-

anitagov commented:
bors r+
